lib/livebook_web/live/session_live/app_info_component.ex

```diff
@@ -0,0 +1,247 @@
+  end
+
+  defp status(%{status: :booting} = assigns) do
+    ~H"""
+    <.status_indicator text="Booting" circle_class="bg-blue-500" animated_circle_class="bg-blue-400" />
+    """
+  end
+
+  defp status(%{status: :running} = assigns) do
+    ~H"""
+    <.status_indicator text="Running" circle_class="bg-green-bright-400" />
+    """
+  end
+
+  defp status(%{status: :error} = assigns) do
+    ~H"""
+    <.status_indicator text="Error" circle_class="bg-red-400" />
+    """
+  end
+
+  defp status(%{status: :shutting_down} = assigns) do
+    ~H"""
+    <.status_indicator text="Shutting down" circle_class="bg-gray-500" />
+    """
+  end
+
+  defp status_indicator(assigns) do
+    assigns =
+      assigns
+      |> assign_new(:animated_circle_class, fn -> nil end)
+
+    ~H"""
+    <div class="flex items-center space-x-2">
+      <div><%= @text %></div>
+      <span class="relative flex h-3 w-3">
+        <%= if @animated_circle_class do %>
+          <span class={"#{@animated_circle_class} animate-ping absolute inline-flex h-full w-full rounded-full opacity-75"}>
+          </span>
+        <% end %>
+        <span class={"#{@circle_class} relative inline-flex rounded-full h-3 w-3 bg-blue-500"}></span>
+      </span>
+    </div>
+    """
+  end
+
+  @impl true
+  def handle_event("validate", %{"app_settings" => params}, socket) do
+    changeset = Livebook.Notebook.AppSettings.change(socket.assigns.settings, params)
+
+    with {:ok, settings} <- Livebook.Notebook.AppSettings.update(socket.assigns.settings, params) do
+      Livebook.Session.set_app_settings(socket.assigns.session.pid, settings)
+    end
+
+    {:noreply, assign(socket, changeset: changeset)}
+  end
+
+  def handle_event("deploy", %{"app_settings" => params}, socket) do
+    case Livebook.Notebook.AppSettings.update(socket.assigns.settings, params) do
+      {:ok, settings} ->
+        Livebook.Session.set_app_settings(socket.assigns.session.pid, settings)
+
+        if slug_taken?(settings.slug, socket.assigns.apps) do
+          {:noreply, assign(socket, deploy_confirmation: true)}
+        else
+          Livebook.Session.deploy_app(socket.assigns.session.pid)
+          {:noreply, socket}
+        end
+
+      {:error, _changeset} ->
+        {:noreply, socket}
+    end
+  end
+
+  def handle_event("deploy_confirmation_confirm", %{}, socket) do
+    Livebook.Session.deploy_app(socket.assigns.session.pid)
+    {:noreply, assign(socket, deploy_confirmation: false)}
+  end
+
+  def handle_event("deploy_confirmation_cancel", %{}, socket) do
+    {:noreply, assign(socket, deploy_confirmation: false)}
+  end
+
+  def handle_event("shutdown_app", %{"session_id" => session_id}, socket) do
+    app = Enum.find(socket.assigns.apps, &(&1.session_id == session_id))
+    Livebook.Session.close(app.session_pid)
+    {:noreply, socket}
+  end
+
+  defp slug_taken?(slug, apps) do
+    own? =
+      Enum.any?(apps, fn app ->
+        app.registered and app.settings.slug == slug
+      end)
```

Comment:
Is this an optimization to avoid going to global if it is registered locally?

Reply:
Nah, it's so that we don't show the confirmation when we are redeploying an app we own. If the app is now shown within this session we show a confirmation "Do you want to replace the app?".